ClientLockWithTerminationTest.testLockOnClient_withNodeCrash

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-pr-builder/com.hazelcast$hazelcast-client/10092/testReport/junit/com.hazelcast.client.lock/ClientLockWithTerminationTest/testLockOnClient_withNodeCrash/

```
Error Message

Lock was obtained by 2 different clients 
Stacktrace

java.lang.AssertionError: Lock was obtained by 2 different clients 
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at org.junit.Assert.assertFalse(Assert.java:64)
    at com.hazelcast.client.lock.ClientLockWithTerminationTest.testLockOnClient_withNodeCrash(ClientLockWithTerminationTest.java:87)
```

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/ProxyManager.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/ProxyManager.java
@@ -24,7 +24,6 @@ import com.hazelcast.client.LoadBalancer;
 import com.hazelcast.client.cache.impl.ClientCacheProxyFactory;
 import com.hazelcast.client.config.ClientConfig;
 import com.hazelcast.client.config.ProxyFactoryConfig;
-import com.hazelcast.client.connection.ClientConnectionManager;
 import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.impl.protocol.codec.ClientAddDistributedObjectListenerCodec;
@@ -78,7 +77,6 @@ import com.hazelcast.mapreduce.impl.MapReduceService;
 import com.hazelcast.multimap.impl.MultiMapService;
 import com.hazelcast.nio.Address;
 import com.hazelcast.nio.ClassLoaderUtil;
-import com.hazelcast.nio.Connection;
 import com.hazelcast.replicatedmap.impl.ReplicatedMapService;
 import com.hazelcast.ringbuffer.impl.RingbufferService;
 import com.hazelcast.scheduledexecutor.impl.DistributedScheduledExecutorService;
@@ -347,35 +345,15 @@ public final class ProxyManager {
     }
 
     private void initialize(ClientProxy clientProxy) throws Exception {
-        final Address initializationTarget = findNextAddressToSendCreateRequest();
-        final Connection connection = getTargetOrOwnerConnection(initializationTarget);
-        final ClientMessage clientMessage = ClientCreateProxyCodec.encodeRequest(clientProxy.getDistributedObjectName(),
+        Address initializationTarget = findNextAddressToSendCreateRequest();
+        if (initializationTarget == null) {
+            throw new IOException("Not able to find a member to create proxy on!");
+        }
+        ClientMessage clientMessage = ClientCreateProxyCodec.encodeRequest(clientProxy.getDistributedObjectName(),
                 clientProxy.getServiceName(), initializationTarget);
-        new ClientInvocation(client, clientMessage, connection).invoke().get();
+        new ClientInvocation(client, clientMessage, initializationTarget).invoke().get();
         clientProxy.setContext(context);
         clientProxy.onInitialize();
-    }
-
-    private Connection getTargetOrOwnerConnection(final Address target) throws IOException {
-        if (target == null) {
-            throw new IOException("Not able to setup owner connection!");
-        }
-
-        final ClientConnectionManager connectionManager = client.getConnectionManager();
-        Connection connection = connectionManager.getConnection(target);
-        if (connection == null) {
-            final Address ownerConnectionAddress = client.getClientClusterService().getOwnerConnectionAddress();
-            if (ownerConnectionAddress == null) {
-                throw new IOException("Not able to setup owner connection!");
-            }
-
-            connection = connectionManager.getConnection(ownerConnectionAddress);
-            if (connection == null) {
-                throw new IOException("Client is not connected to member " + target);
-            }
-        }
-
-        return connection;
     }
 
     public Address findNextAddressToSendCreateRequest() {

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java
@@ -20,17 +20,19 @@ import com.hazelcast.client.HazelcastClientNotActiveException;
 import com.hazelcast.client.connection.nio.ClientConnection;
 import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.spi.ClientClusterService;
 import com.hazelcast.client.spi.ClientExecutionService;
 import com.hazelcast.client.spi.EventHandler;
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
 import com.hazelcast.core.HazelcastOverloadException;
 import com.hazelcast.core.LifecycleService;
-import com.hazelcast.logging.ILogger;
 import com.hazelcast.core.OperationTimeoutException;
+import com.hazelcast.logging.ILogger;
 import com.hazelcast.nio.Address;
 import com.hazelcast.nio.Connection;
 import com.hazelcast.spi.exception.RetryableException;
 import com.hazelcast.spi.exception.TargetDisconnectedException;
+import com.hazelcast.spi.exception.TargetNotMemberException;
 
 import java.io.IOException;
 import java.util.concurrent.Executor;
@@ -53,6 +55,7 @@ public class ClientInvocation implements Runnable {
     private final ClientInvocationFuture clientInvocationFuture;
     private final ILogger logger;
     private final LifecycleService lifecycleService;
+    private final ClientClusterService clientClusterService;
     private final ClientInvocationServiceSupport invocationService;
     private final ClientExecutionService executionService;
     private final ClientMessage clientMessage;
@@ -70,6 +73,7 @@ public class ClientInvocation implements Runnable {
                                int partitionId,
                                Address address,
                                Connection connection) {
+        this.clientClusterService = client.getClientClusterService();
         this.lifecycleService = client.getLifecycleService();
         this.invocationService = (ClientInvocationServiceSupport) client.getInvocationService();
         this.executionService = client.getClientExecutionService();
@@ -177,7 +181,7 @@ public class ClientInvocation implements Runnable {
             return;
         }
 
-        if (isBindToSingleConnection() && exception instanceof IOException) {
+        if (isNotAllowedToRetryOnSelection(exception)) {
             clientInvocationFuture.complete(exception);
             return;
         }
@@ -206,6 +210,23 @@ public class ClientInvocation implements Runnable {
             clientInvocationFuture.complete(exception);
         }
 
+    }
+
+    private boolean isNotAllowedToRetryOnSelection(Throwable exception) {
+        if (isBindToSingleConnection() && exception instanceof IOException) {
+            return true;
+        }
+
+        if (address != null
+                && exception instanceof TargetNotMemberException
+                && clientClusterService.getMember(address) == null) {
+            //when invocation send over address
+            //if exception is target not member and
+            //address is not available in member list , don't retry
+            clientInvocationFuture.complete(exception);
+            return true;
+        }
+        return false;
     }
 
     private boolean isBindToSingleConnection() {
@@ -267,7 +288,6 @@ public class ClientInvocation implements Runnable {
         }
         return "ClientInvocation{"
                 + "clientMessageType=" + clientMessage.getMessageType()
-                + ", callIdSequence=" + callIdSequence
                 + ", target=" + target
                 + ", sendConnection=" + sendConnection + '}';
     }

